# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [2]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [12]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head(20)

PassengerId HomePlanet CryoSleep  Cabin    Destination   Age    VIP  \
0      0001_01     Europa     False  B/0/P    TRAPPIST-1e  39.0  False   
1      0002_01      Earth     False  F/0/S    TRAPPIST-1e  24.0  False   
2      0003_01     Europa     False  A/0/S    TRAPPIST-1e  58.0   True   
3      0003_02     Europa     False  A/0/S    TRAPPIST-1e  33.0  False   
4      0004_01      Earth     False  F/1/S    TRAPPIST-1e  16.0  False   
5      0005_01      Earth     False  F/0/P  PSO J318.5-22  44.0  False   
6      0006_01      Earth     False  F/2/S    TRAPPIST-1e  26.0  False   
7      0006_02      Earth      True  G/0/S    TRAPPIST-1e  28.0  False   
8      0007_01      Earth     False  F/3/S    TRAPPIST-1e  35.0  False   
9      0008_01     Europa      True  B/1/P    55 Cancri e  14.0  False   
10     0008_02     Europa      True  B/1/P    TRAPPIST-1e  34.0  False   
11     0008_03     Europa     False  B/1/P    55 Cancri e  45.0  False   
12     0009_01       Mars     False  F/1/P    TRAPPIST-1e  32.0  False   
13     0010_01      Earth     False  G/1/S    TRAPPIST-1e  48.0  False   
14     0011_01      Earth     False  F/2/P    TRAPPIST-1e  28.0  False   
15     0012_01      Earth     False    NaN    TRAPPIST-1e  31.0  False   
16     0014_01       Mars     False  F/3/P    55 Cancri e  27.0  False   
17     0015_01      Earth     False  F/4/P    55 Cancri e  24.0  False   
18     0016_01       Mars      True  F/5/P    TRAPPIST-1e  45.0  False   
19     0017_01      Earth     False  G/0/P    TRAPPIST-1e   0.0  False   

    RoomService  FoodCourt  ShoppingMall     Spa  VRDeck                Name  \
0           0.0        0.0           0.0     0.0     0.0     Maham Ofracculy   
1         109.0        9.0          25.0   549.0    44.0        Juanna Vines   
2          43.0     3576.0           0.0  6715.0    49.0       Altark Susent   
3           0.0     1283.0         371.0  3329.0   193.0        Solam Susent   
4         303.0       70.0         151.0   565.0     2.0   Willy Santantines   
5           0.0      483.0           0.0   291.0     0.0   Sandie Hinetthews   
6          42.0     1539.0           3.0     0.0     0.0  Billex Jacostaffey   
7           0.0        0.0           0.0     0.0     NaN  Candra Jacostaffey   
8           0.0      785.0          17.0   216.0     0.0       Andona Beston   
9           0.0        0.0           0.0     0.0     0.0      Erraiam Flatic   
10          0.0        0.0           NaN     0.0     0.0      Altardr Flatic   
11         39.0     7295.0         589.0   110.0   124.0       Wezena Flatic   
12         73.0        0.0        1123.0     0.0   113.0        Berers Barne   
13        719.0        1.0          65.0     0.0    24.0      Reney Baketton   
14          8.0      974.0          12.0     2.0     7.0     Elle Bertsontry   
15         32.0        0.0         876.0     0.0     0.0       Justie Pooles   
16       1286.0      122.0           NaN     0.0     0.0         Flats Eccle   
17          0.0        1.0           0.0     0.0   637.0     Carry Hughriend   
18          0.0        0.0           0.0     0.0     0.0          Alus Upead   
19          0.0        0.0           0.0     0.0     0.0       Lyde Brighttt   

    Transported  
0         False  
1          True  
2         False  
3         False  
4          True  
5          True  
6          True  
7          True  
8          True  
9          True  
10         True  
11         True  
12         True  
13        False  
14         True  
15        False  
16        False  
17        False  
18         True  
19         True

**Check the shape of your data**

In [4]:
spaceship.shape

(8693, 14)

**Check for data types**

In [5]:
spaceship.dtypes

PassengerId      object
HomePlanet       object
CryoSleep        object
Cabin            object
Destination      object
Age             float64
VIP              object
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
Name             object
Transported        bool
dtype: object

**Check for missing values**

In [6]:
spaceship.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continouos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value. 

In [9]:
spaceship_cleaned = spaceship.dropna()

In [10]:
spaceship_cleaned.isnull().sum()

PassengerId     0
HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
Transported     0
dtype: int64

In [14]:
spaceship_cleaned.shape

(6606, 14)

In [15]:
spaceship_cleaned.head(10)

PassengerId HomePlanet CryoSleep  Cabin    Destination   Age    VIP  \
0      0001_01     Europa     False  B/0/P    TRAPPIST-1e  39.0  False   
1      0002_01      Earth     False  F/0/S    TRAPPIST-1e  24.0  False   
2      0003_01     Europa     False  A/0/S    TRAPPIST-1e  58.0   True   
3      0003_02     Europa     False  A/0/S    TRAPPIST-1e  33.0  False   
4      0004_01      Earth     False  F/1/S    TRAPPIST-1e  16.0  False   
5      0005_01      Earth     False  F/0/P  PSO J318.5-22  44.0  False   
6      0006_01      Earth     False  F/2/S    TRAPPIST-1e  26.0  False   
8      0007_01      Earth     False  F/3/S    TRAPPIST-1e  35.0  False   
9      0008_01     Europa      True  B/1/P    55 Cancri e  14.0  False   
11     0008_03     Europa     False  B/1/P    55 Cancri e  45.0  False   

    RoomService  FoodCourt  ShoppingMall     Spa  VRDeck                Name  \
0           0.0        0.0           0.0     0.0     0.0     Maham Ofracculy   
1         109.0        9.0          25.0   549.0    44.0        Juanna Vines   
2          43.0     3576.0           0.0  6715.0    49.0       Altark Susent   
3           0.0     1283.0         371.0  3329.0   193.0        Solam Susent   
4         303.0       70.0         151.0   565.0     2.0   Willy Santantines   
5           0.0      483.0           0.0   291.0     0.0   Sandie Hinetthews   
6          42.0     1539.0           3.0     0.0     0.0  Billex Jacostaffey   
8           0.0      785.0          17.0   216.0     0.0       Andona Beston   
9           0.0        0.0           0.0     0.0     0.0      Erraiam Flatic   
11         39.0     7295.0         589.0   110.0   124.0       Wezena Flatic   

    Transported  
0         False  
1          True  
2         False  
3         False  
4          True  
5          True  
6          True  
8          True  
9          True  
11         True

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [18]:
spaceship_cleaned['Cabin'] = spaceship_cleaned['Cabin'].apply(lambda x: x[0])

display(spaceship_cleaned['Cabin'].unique())

array(['B', 'F', 'A', 'G', 'E', 'C', 'D', 'T'], dtype=object)

In [19]:
display(spaceship_cleaned.head())

PassengerId HomePlanet CryoSleep Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False     B  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False     F  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False     A  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False     A  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False     F  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

- Drop PassengerId and Name

In [21]:
spaceship_cleaned_2 = spaceship_cleaned.drop(columns=['PassengerId', 'Name'])
display(spaceship_cleaned_2.head())

HomePlanet CryoSleep Cabin  Destination   Age    VIP  RoomService  \
0     Europa     False     B  TRAPPIST-1e  39.0  False          0.0   
1      Earth     False     F  TRAPPIST-1e  24.0  False        109.0   
2     Europa     False     A  TRAPPIST-1e  58.0   True         43.0   
3     Europa     False     A  TRAPPIST-1e  33.0  False          0.0   
4      Earth     False     F  TRAPPIST-1e  16.0  False        303.0   

   FoodCourt  ShoppingMall     Spa  VRDeck  Transported  
0        0.0           0.0     0.0     0.0        False  
1        9.0          25.0   549.0    44.0         True  
2     3576.0           0.0  6715.0    49.0        False  
3     1283.0         371.0  3329.0   193.0        False  
4       70.0         151.0   565.0     2.0         True

- For non-numerical columns, do dummies.

In [24]:
spaceship_cleaned_2 = pd.get_dummies(spaceship_cleaned_2)
boolean_columns = spaceship_cleaned_2.select_dtypes(include=['bool']).columns
spaceship_cleaned_2[boolean_columns] = spaceship_cleaned_2[boolean_columns].astype(int)

display(spaceship_cleaned_2.head())

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0  39.0          0.0        0.0           0.0     0.0     0.0            0   
1  24.0        109.0        9.0          25.0   549.0    44.0            1   
2  58.0         43.0     3576.0           0.0  6715.0    49.0            0   
3  33.0          0.0     1283.0         371.0  3329.0   193.0            0   
4  16.0        303.0       70.0         151.0   565.0     2.0            1   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_D  \
0                 0                  1                0  ...        0   
1                 1                  0                0  ...        0   
2                 0                  1                0  ...        0   
3                 0                  1                0  ...        0   
4                 1                  0                0  ...        0   

   Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0        0        0        0        0                        0   
1        0        1        0        0                        0   
2        0        0        0        0                        0   
3        0        0        0        0                        0   
4        0        1        0        0                        0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                          0                        1          1         0  
1                          0                        1          1         0  
2                          0                        1          0         1  
3                          0                        1          1         0  
4                          0                        1          1         0  

[5 rows x 25 columns]

**Perform Train Test Split**

In [25]:
from sklearn.model_selection import train_test_split

# Perform train-test split
X = spaceship_cleaned_2.drop(columns=['Transported'])  # Features (assuming 'Transported' is the target variable)
y = spaceship_cleaned_2['Transported']  # Target variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Verify the shapes of the resulting datasets
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (5284, 24)
X_test shape: (1322, 24)
y_train shape: (5284,)
y_test shape: (1322,)


**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [29]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Initialize and train the KNN model
knn = KNeighborsClassifier(n_neighbors=5) 
knn.fit(X_train, y_train)

# Make predictions
y_pred = knn.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print the evaluation results
print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")
print(f"Confusion Matrix:\n{conf_matrix}")

Accuracy: 0.7859304084720121
Classification Report:
              precision    recall  f1-score   support

           0       0.78      0.79      0.79       653
           1       0.79      0.78      0.79       669

    accuracy                           0.79      1322
   macro avg       0.79      0.79      0.79      1322
weighted avg       0.79      0.79      0.79      1322

Confusion Matrix:
[[518 135]
 [148 521]]


- Evaluate your model's performance. Comment it

1. Accuracy
The model achieves an accuracy of approximately 78.59%. This means that about 78.59% of the predictions made by the model on the test set were correct.

2. Precision and Recall
2.1 Precision: Precision for class 0 (not transported) is 0.78, and for class 1 (transported) is 0.79. This indicates that when the model predicts a passenger was transported or not transported, it is correct about 78% to 79% of the time.

2.2 Recall: Recall for class 0 is 0.79, and for class 1 is 0.78. This suggests that the model correctly identifies about 79% of the passengers who were actually not transported and about 78% of those who were actually transported.

3. Confusion Matrix
The confusion matrix provides a detailed breakdown of predictions:

True Positives (TP): 521 passengers correctly predicted as transported.
True Negatives (TN): 518 passengers correctly predicted as not transported.
False Positives (FP): 135 passengers incorrectly predicted as transported (Type I error).
False Negatives (FN): 148 passengers incorrectly predicted as not transported (Type II error).

4. Overall Performance:

The model shows balanced performance with similar precision, recall, and F1-score for both classes (transported and not transported). This indicates that the model does not favor one class over the other significantly.